From bb1d577d35d551cfceb8a2c70d9668f65275c6fb Mon Sep 17 00:00:00 2001
From: wszqkzqk <wszqkzqk@qq.com>
Date: Sat, 1 Oct 2022 08:37:31 +0800
Subject: [PATCH] Optional optimization for `g_int64_hash`

---
 glib/ghash.c | 2 +-
 1 file changed, 1 insertion(+), 1 deletion(-)

diff --git a/glib/ghash.c b/glib/ghash.c
index 4077ad374..429f38d63 100644
--- a/glib/ghash.c
+++ b/glib/ghash.c
@@ -2490,7 +2490,7 @@ g_int64_equal (gconstpointer v1,
 guint
 g_int64_hash (gconstpointer v)
 {
-  return (guint) *(const gint64*) v;
+  return (guint) ((const guint) (*(guint64*) v >> 32)) ^ (*(const guint*) v);
 }
 
 /**
-- 
2.38.0.windows.1

